_includes/tutorials/window-final-result/kstreams/code/src/test/java/io/confluent/developer/WindowFinalResultTest.java

```diff
@@ -0,0 +1,191 @@
+import org.apache.kafka.streams.TopologyTestDriver;
+import org.apache.kafka.streams.kstream.TimeWindows;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.test.ConsumerRecordFactory;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.*;
+
+import static com.github.grantwest.eventually.EventuallyLambdaMatcher.eventuallyEval;
+import static io.confluent.kafka.serializers.AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG;
+import static org.apache.kafka.streams.kstream.WindowedSerdes.timeWindowedSerdeFrom;
+import static org.hamcrest.core.Is.is;
+
+public class WindowFinalResultTest {
+
+    private TopologyTestDriver topologyTestDriver;
+    private SpecificAvroSerde<PressureAlert> pressureSerde;
+
+    private Config config = ConfigFactory.load("test.properties");
+
+    private String inputTopic = this.config.getString("input.topic.name");
+    private String outputTopic = this.config.getString("output.topic.name");
+
+    private Duration testWindowSize = config.getDuration("window.size");
+    private Duration testGracePeriodSize = config.getDuration("window.grace.period");
+    private Serde<Windowed<String>> keyResultSerde = timeWindowedSerdeFrom(String.class, testWindowSize.toMillis());
+
+    private TimeWindows makeFixedTimeWindow() {
+        return TimeWindows.of(testWindowSize).advanceBy(testWindowSize).grace(testGracePeriodSize);
+    }
+
+    private SpecificAvroSerde<PressureAlert> makePressureAlertSerde() throws IOException, RestClientException {
+
+        final MockSchemaRegistryClient client = new MockSchemaRegistryClient();
+
+        String subject = String.format("%s-value", config.getString("input.topic.name"));
+
+        client.register(subject, PressureAlert.SCHEMA$);
+
+        Map<String, String> schemaRegistryConfigMap = Collections.singletonMap(
+                SCHEMA_REGISTRY_URL_CONFIG,
+                config.getString(SCHEMA_REGISTRY_URL_CONFIG)
+        );
+
+        SpecificAvroSerde<PressureAlert> serde = new SpecificAvroSerde<>(client);
+
+        serde.configure(schemaRegistryConfigMap, false);
+
+        return serde;
+    }
+
+    private ProducerRecord<Windowed<String>, Long> readNext() {
+        return topologyTestDriver.readOutput(
+                this.outputTopic,
+                this.keyResultSerde.deserializer(),
+                Serdes.Long().deserializer()
+        );
+    }
+
+    private List<ProducerRecord<Windowed<String>, Long>> readAtLeastNOutputs(int size) {
+        List<ProducerRecord<Windowed<String>, Long>> result = new ArrayList<>();
+
+        Assert.assertThat(() -> {
+            ProducerRecord<Windowed<String>, Long> record = readNext();
+            if(null != record) result.add(record);
+            return result.size();
+        }, eventuallyEval(is(size), this.testGracePeriodSize));
```

Comment:
I'm having second thoughts about this. Since the only way to move the streaming time and get my aggregation result is to send a new event, I will get my result immediately anyway. It works fine as long as the last piped record has a timestamp greater than `window size + grace period`. There is no need to wait here 🤔.

Reply:
I think it makes sense to leave it this way. Some tutorial users may experiment with test data that does not meet the window size + grace period criteria, causing the test to unexpectedly fail, if this logic were simplified.

Reply:
I think, in this case, users will also have to send a final event (the same way I do at line 115 and 165) to get back their test data. Without that, tests will fail no matter what. What do you think?